selfservice/strategy/code/strategy.go

```diff
@@ -82,14 +119,348 @@ func NewStrategy(deps strategyDependencies) *Strategy {
+		}
+	}
+
+	return nil
+}
+
+type (
+	FlowType interface {
+		*login.Flow | *registration.Flow | *recovery.Flow | *verification.Flow
+		flow.Flow
+	}
+	FlowPayload interface {
+		*updateLoginFlowWithCodeMethod | *updateRegistrationFlowWithCodeMethod | *updateRecoveryFlowWithCodeMethod | *updateVerificationFlowWithCodeMethod
+	}
+	CreateCodeState[T FlowType, P FlowPayload]           func(context.Context, T, *Strategy, P) error
+	ValidateCodeState[T FlowType, P FlowPayload]         func(context.Context, T, *Strategy, P) error
+	AlreadyValidatedCodeState[T FlowType, P FlowPayload] func(context.Context, T, *Strategy, P) error
+	CodeStateManager[T FlowType, P FlowPayload]          struct {
+		f                         T
+		payload                   P
+		strategy                  *Strategy
+		createCodeState           CreateCodeState[T, P]
+		verifyCodeState           ValidateCodeState[T, P]
+		alreadyValidatedCodeState AlreadyValidatedCodeState[T, P]
+	}
+)
+
+func NewCodeStateManager[T FlowType, P FlowPayload](f T, s *Strategy, payload P) *CodeStateManager[T, P] {
+	return &CodeStateManager[T, P]{
+		f:        f,
+		strategy: s,
+		payload:  payload,
+	}
+}
+
+func (c *CodeStateManager[T, P]) SetCreateCodeHandler(fn CreateCodeState[T, P]) {
+	c.createCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) SetCodeVerifyHandler(fn ValidateCodeState[T, P]) {
+	c.verifyCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) SetCodeDoneHandler(fn AlreadyValidatedCodeState[T, P]) {
+	c.alreadyValidatedCodeState = fn
+}
+
+func (c *CodeStateManager[T, P]) validatePayload(ctx context.Context) error {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		if len(v.Identifier) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/identifier", "identifier"))
+		}
+	case *updateRegistrationFlowWithCodeMethod:
+		if len(v.Traits) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/traits", "traits"))
+		}
+	case *updateRecoveryFlowWithCodeMethod:
+		if len(v.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	case *updateVerificationFlowWithCodeMethod:
+		if len(v.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	default:
+		return errors.WithStack(herodot.ErrBadRequest.WithReason("received unexpected flow payload type"))
+	}
+	return nil
+}
+
+func (c *CodeStateManager[T, P]) getResend() string {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		return v.Resend
+	case *updateRegistrationFlowWithCodeMethod:
+		return v.Resend
+	}
+	return ""
+}
+
+func (c *CodeStateManager[T, P]) getCode() string {
+	switch v := any(c.payload).(type) {
+	case *updateLoginFlowWithCodeMethod:
+		return v.Code
+	case *updateRegistrationFlowWithCodeMethod:
+		return v.Code
+	case *updateRecoveryFlowWithCodeMethod:
+		return v.Code
+	case *updateVerificationFlowWithCodeMethod:
+		return v.Code
+	}
+	return ""
+}
+
```

Comment:
This should be interfaces that implement `ShouldResend` and `GetCode` for example, no need for type assertions. It makes the code more difficult to read and navigate and since Go has duck typing, also not needed.